server-cloud/server/auth.go

```diff
@@ -0,0 +1,199 @@
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+	}
+
+	state, err := generateRandomState()
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+	}
+
+	sess, err := session.Get("secret", c)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+	}
+
+	sess.Values["state"] = state
+
+	if err := sess.Save(c.Request(), c.Response()); err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+	}
+
+	return c.Redirect(http.StatusTemporaryRedirect, auth.AuthCodeURL(state))
+}
+
+func (s *Server) callback(c echo.Context) error {
+	auth, err := getAuth(c.Request().Context(), s.conf)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+	}
+
+	sess, err := session.Get("secret", c)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+	}
+
+	// session := sessions.Default(ctx)
+	if c.QueryParam("state") != sess.Values["state"] {
+		return c.String(http.StatusBadRequest, "Invalid state parameter.")
+	}
+
+	// Exchange an authorization code for a token.
+	token, err := auth.Exchange(c.Request().Context(), c.QueryParam("code"))
+	if err != nil {
+		return c.String(http.StatusUnauthorized, "Failed to convert an authorization code into a token.")
+	}
+
+	idToken, err := auth.VerifyIDToken(c.Request().Context(), token)
+	if err != nil {
+		return c.String(http.StatusInternalServerError, "Failed to verify ID Token.")
+	}
+
+	var profile map[string]interface{}
+	if err := idToken.Claims(&profile); err != nil {
+		c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	sess.Values["access_token"] = token.AccessToken
+	sess.Values["profile"] = profile
+
+	if err := sess.Save(c.Request(), c.Response()); err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	// Redirect to logged in page.
+	c.Redirect(http.StatusTemporaryRedirect, "/user")
+	return err
+}
+
+func (s *Server) logout(c echo.Context) error {
+	logoutUrl, err := url.Parse("https://" + s.conf.Auth0Domain + "/v2/logout")
+	if err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	scheme := "http"
+	if c.Request().TLS != nil {
+		scheme = "https"
+	}
+
+	fmt.Println(c.Request().Host)
+	returnTo, err := url.Parse(scheme + "://" + c.Request().Host)
+	if err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+
+	}
+
+	parameters := url.Values{}
+	parameters.Add("returnTo", returnTo.String())
+	parameters.Add("client_id", s.conf.Auth0ClientID)
+	logoutUrl.RawQuery = parameters.Encode()
+
+	return c.Redirect(http.StatusTemporaryRedirect, logoutUrl.String())
+}
+
+func (s *Server) user(c echo.Context) error {
+	sess, err := session.Get("secret", c)
+	if err != nil {
+		fmt.Print(err)
+	}
+
+	profile := sess.Values["profile"]
+	// return c.HTML(http.StatusOK, "profile")
+	return c.Render(http.StatusOK, "user.html", profile)
+}
```

Comment:
This should return the profile directly as a JSON object – not by rendering a template.

Reply:
Ohk, just added for testing it using html templated in the sample code, will update this 👍